File: platforms/ios/Packages/FunputKit/Sources/KeyboardRenderer/Clipboard/ClipboardRowText.swift
import Foundation

/// Turns a stored clip into the two strings a row shows.
///
/// The relative time is built by hand rather than with `RelativeDateTimeFormatter`:
/// the wording then cannot drift with OS locale data, a test can pin it exactly, and
/// no formatter has to be allocated or kept alive per row.
enum ClipboardRowText {
    static let previewLimit = 120

    /// Collapses every run of whitespace — newlines included — into one space.
    ///
    /// The row shows at most two lines, so a clip that starts with blank lines would
    /// otherwise render as an empty row. Structure is lost, but the point here is to
    /// recognise the clip, not to read it.
    static func preview(_ text: String) -> String {
        let collapsed = text
            .split(whereSeparator: \.isWhitespace)
            .joined(separator: " ")
        guard collapsed.count > previewLimit else { return collapsed }
        return collapsed.prefix(previewLimit) + "…"
    }

    static func relativeTime(from date: Date, now: Date = Date()) -> String {
        let seconds = Int(now.timeIntervalSince(date))
        guard seconds >= 60 else { return "Vừa xong" }
        let minutes = seconds / 60
        guard minutes >= 60 else { return "\(minutes) phút trước" }
        let hours = minutes / 60
        guard hours >= 24 else { return "\(hours) giờ trước" }
        return "\(hours / 24) ngày trước"
    }
}

File: platforms/android/keyboard-ui/src/main/java/app/funput/funput/keyboard/ui/clipboard/ClipboardRowText.kt
package app.funput.funput.keyboard.ui.clipboard

import java.time.Duration
import java.time.Instant

internal object ClipboardRowText {
    const val PreviewLimit = 120

    fun preview(text: String): String {
        val collapsed = text.trim().replace(Whitespace, " ")
        val count = collapsed.codePointCount(0, collapsed.length)
        if (count <= PreviewLimit) return collapsed
        val end = collapsed.offsetByCodePoints(0, PreviewLimit)
        return collapsed.substring(0, end) + "…"
    }

    fun relativeTime(
        capturedAt: Instant,
        now: Instant,
        strings: ClipboardTimeStrings,
    ): String {
        val seconds = Duration.between(capturedAt, now).seconds.coerceAtLeast(0)
        if (seconds < 60) return strings.justNow
        val minutes = seconds / 60
        if (minutes < 60) return strings.minutesAgo(minutes)
        val hours = minutes / 60
        return if (hours < 24) strings.hoursAgo(hours) else strings.daysAgo(hours / 24)
    }

    private val Whitespace = Regex("\\s+")
}

internal data class ClipboardTimeStrings(
    val justNow: String,
    val minutesAgo: (Long) -> String,
    val hoursAgo: (Long) -> String,
    val daysAgo: (Long) -> String,
)
